pkg/workspaceapi/api.go

```diff
@@ -0,0 +1,178 @@
+	ECodeRpcSerialization  = "warpforge-error-rpc-serialization"
+	ECodeRpcUnknown        = "warpforge-error-rpc-unknown"
+	ECodeRpcConnection     = "warpforge-error-rpc-connection"
+	ECodeRpcMissingData    = "warpforge-error-rpc-missing-data"
+	ECodeRpcExtraData      = "warpforge-error-rpc-extra-data"
+)
+
+const (
+	ModuleStatus_NoInfo             ModuleStatus = "NoInfo"
+	ModuleStatus_Queuing            ModuleStatus = "Queuing"
+	ModuleStatus_InProgress         ModuleStatus = "InProgress"
+	ModuleStatus_FailedProvisioning ModuleStatus = "FailedProvisioning"
+	ModuleStatus_ExecutedSuccess    ModuleStatus = "ExecutedSuccess"
+	ModuleStatus_ExecutedFailed     ModuleStatus = "ExecutedFailed"
+)
+
+type RpcRequest struct {
+	*ModuleStatusQuery
+}
+
+type RpcResponse struct {
+	*Echo
+	*ModuleStatusAnswer
+	*Error
+}
+
+type RpcData struct {
+	*RpcRequest
+	*RpcResponse
+}
+
+type Rpc struct {
+	ID   string
+	Data RpcData
+}
+
+type Error struct {
+	Code    string
+	Message *string
+	Details *Details
+	Cause   *Error
+}
+
+type Details struct {
+	Keys   []string
+	Values map[string]string
+}
+
+func (d *Details) Details() [][2]string {
+	if d == nil {
+		return [][2]string{}
+	}
+	result := make([][2]string, 0, len(d.Keys))
+	for _, key := range d.Keys {
+		result = append(result, [2]string{key, d.Values[key]})
+	}
+	return result
+}
+
+func (e *Error) serum() *serum.ErrorValue {
+	data := e.AsSerumData()
+	if data == nil {
+		return nil
+	}
+	return &serum.ErrorValue{Data: *data}
+}
+
+func (e *Error) AsSerumData() *serum.Data {
+	if e == nil {
+		return nil
+	}
+	var msg string
+	if e.Message != nil {
+		msg = *e.Message
+	}
+	return &serum.Data{
+		Code:    e.Code,
+		Message: msg,
+		Details: e.Details.Details(),
+		Cause:   e.Cause.serum(),
+	}
+}
+
+func bindnodeCopy(data datamodel.Node, i interface{}, t schema.Type) (interface{}, error) {
+	np := bindnode.Prototype(i, t)
+	nb := np.NewBuilder()
+	if err := datamodel.Copy(data, nb); err != nil {
+		return nil, err
+	}
+	result := bindnode.Unwrap(nb.Build())
+	return result, nil
+}
+
+type ModuleStatusQuery struct {
+	Path              string
+	InputReplacements *InputReplacements
+	InterestLevel     ModuleInterestLevel
+}
+
+type InputReplacements struct {
+	Keys   []wfapi.PlotInput
+	Values map[wfapi.PlotInput]wfapi.WareID
+}
+
+type ModuleInterestLevel string
+
+type ModuleStatus string
+
+type ModuleStatusAnswer struct {
+	Path   string
+	Status ModuleStatus
+}
+
+type Echo string
+
+type ModuleStatusUnion struct {
+	ModuleStatusUnion_NoInfo             *ModuleStatusUnion_NoInfo
+	ModuleStatusUnion_Queuing            *ModuleStatusUnion_Queuing
+	ModuleStatusUnion_InProgress         *ModuleStatusUnion_InProgress
+	ModuleStatusUnion_FailedProvisioning *ModuleStatusUnion_FailedProvisioning
+	ModuleStatusUnion_ExecutedSuccess    *ModuleStatusUnion_ExecutedSuccess
+	ModuleStatusUnion_ExecutedFailed     *ModuleStatusUnion_ExecutedFailed
+}
+
+func UnionField(i interface{}) string {
```

Comment:
You can also just get this by looking at it with IPLD stuff and asking the map key.

Unions when viewed at type level always act like a map with one key, and the key is the name of the member type.

Reply:
Positive: less reflect (or just hiding it behind the IPLD lib). Negative: a bit weird. Maybe there's a better way to do it that I'm not seeing.
 
```golang
// Errors:
//
//   - warpforge-error-unknown -- Unable to extract union key
func (r *RpcRequest) Kind() (string, error) {
	sch := TypeSystem.TypeByName("RpcRequest")
	return unionField(r, sch)
}

// Errors:
//
//   - warpforge-error-unknown -- Unable to extract union key
func unionField(i interface{}, sch schema.Type) (string, error) {
	n := bindnode.Wrap(i, sch)
	iter := n.MapIterator()
	key, _, err := iter.Next()
	if err != nil {
		return "", serum.Error(wfapi.ECodeUnknown, serum.WithCause(err))
	}
	result, err := key.AsString()
	if err != nil {
		return "", serum.Error(wfapi.ECodeUnknown, serum.WithCause(err))
	}
	return result, nil
}
```